src/effects/How to make an effect stateless.txt

```diff
@@ -0,0 +1,189 @@
+4 STEPS TO MAKE AN EFFECT STATELESS
+
+1) "Define EffectEchoSettings"
+
+   In header file:
+       before the effect definition, define a struct EffectEchoSettings containing the settings.
+
+       Base class remains: StatefulPerTrackEffect
+       
+       Declare a private EffectEchoSettings mSettings;
+       
+       REMOVE the members of the effect that go into the settings. REPLACE instead with EffectEchoSettings mSettings; in the effect and then do what is needed to fix compilation and running:
+          
+       Modify FetchParameters: it must now return the address of mSettings, not of the effect
+    
+       Update pointers-to-members in EffectParameter declarations
+    
+   In cpp file:
+       Insert sufficient auto& echoSettings = mSettings; (which changes later) and echoSettings. (which won't need more change) to fix compilation
+        
+       Remove call to Reset() in constructor
+        
+    
+2) "Define validator"
+
+    In header file:
+    
+    add line "struct Validator;" in public section
+
+
+    In cpp file:
+    add class definition in the .cpp with mSettings:
+    
+        struct EffectEcho::Validator
+           : DefaultEffectUIValidator
+        {
+           Validator(EffectUIClientInterface& effect,
+              EffectSettingsAccess& access, EffectEchoSettings& settings)
+              : DefaultEffectUIValidator{ effect, access }
+              , mSettings{ settings }
+           {}
+           virtual ~Validator() = default;
+
+           Effect& GetEffect() const { return static_cast<Effect&>(mEffect); }
+
+           bool ValidateUI() override;
+           bool UpdateUI() override;
+
+           void PopulateOrExchange(ShuttleGui& S);
+
+           EffectEchoSettings& mSettings;
+        };
+
+
+        bool EffectEcho::Validator::ValidateUI()
+        {
+           mAccess.ModifySettings
+           (
+              [this](EffectSettings& settings)
+           {
+              // pass back the modified settings to the MessageBuffer
+
+              // TODO uncomment at last step
+              //EffectEcho::GetSettings(settings) = mSettings;
```

Comment:
I did say, avoid real use of `GetSettings` until the last step.

Another possibility, which I am pursuing in my transformation of AudioUnit effects (a big recent PR that is still a draft and doesn't yet do the validator), is that I MAY put all the needed calls to GetSettings() in early, BUT I do not yet use EffectWithSettings<> and the definition of GetSettings() that it generates.  Instead, I define GetSettings() to ignore its argument and return mSettings in the effect object.  At the last step of real statelessness, I can delete the definition of GetSettings() but leave the calls.